idp/client.ts

```diff
@@ -0,0 +1,135 @@
+import { Context } from "aws-lambda";
+import { GetSecretValueCommandInput } from "@aws-sdk/client-secrets-manager";
+import axios from "axios";
+import { Logger } from "@aws-lambda-powertools/logger";
+import { secretsClient } from "../utils/aws-sdk/secrets-manager";
+
+const logger = new Logger({ serviceName: "TestIdp" });
+
+const IDP_CLIENT_ID = process.env.IDP_CLIENT_ID!;
+const IDP_CLIENT_SECRET_NAME = process.env.IDP_CLIENT_SECRET_NAME!;
+const IDP_DOMAIN = process.env.IDP_DOMAIN!;
+
+/**
+ * Very basic encoding of the Cognito TOKEN endpoint response objects.
+ */
+interface TokenResponse {
+  access_token: string;
+  id_token?: string;
+  refresh_token?: string;
+  expires_in: number;
+  token_type: "Bearer";
+}
+
+/**
+ * Build a client_credentials flow request body
+ *
+ * This does not support requesting any scopes.
+ *
+ * @param clientId The OAuth Client ID
+ * @param scopes A list of scopes to request
+ */
+function buildTokenRequest(clientId: string, scopes?: string[]): URLSearchParams {
+  const form = new URLSearchParams();
+  form.set("grant_type", "client_credentials");
+  form.set("client_id", clientId);
+  if (scopes?.length) {
+    form.set("scope", scopes.join(" "));
+  }
```

Comment:
I am curious why `scopes` are optional for the `buildTokenRequest` function?

Reply:
Great question!! So https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html#post-token-request-parameters-in-body will be a good reference here. By default (at least with Cognito) the server will return all scopes the client is configured for. This allows us to limit the scopes if we want. 

From the Cognito docs:

> Can be a combination of any custom scopes associated with an app client. Any scope that you request must be activated for the app client, or Amazon Cognito will ignore it. If the client doesn't request any scopes, the authentication server uses all custom scopes associated with the client.

Even when using `client_credentials`, this is still an optional argument with a decent default behavior.